CORE-1484 Add "first class" instant launches to navigation drawer

ash3rz opened the pull request:
NOTE: The terrain and k8s PR will need to be merged prior to this one.  Also may require some additional testing once the QA environment is back up.

![image](https://user-images.githubusercontent.com/8909156/135366691-fb7bb62b-d5f2-49e7-a028-a1eb8867eab7.png)

ash3rz commented on src/components/instantlaunches/InstantLaunchButtonWrapper.js:
```diff
@@ -0,0 +1,73 @@
+        <>
+            {render(onClick)}
+            <InstantLaunchSubmissionDialog open={open} />
+        </>
```
I extracted all of the above logic from the `InstantLaunchButton` in `src/components/instantlaunches/index.js` so I could reuse it in the navigation drawer.

sri2k1us commented:
Nirav wants to use 
![image](https://user-images.githubusercontent.com/1250790/135477411-678854ce-a16d-4004-8b9b-dbf8225e6ab0.png)
 for Cloud Shell app. Can we hard-code/ use some constants to get that icon in place for now until we get custom app icon feature implemented ?

sri2k1us commented on config/default.yaml:
```diff
@@ -89,3 +89,6 @@ unleash:
 
 metadata:
     community_attr: cyverse-community
+    instant_launch_location_attr: ui_location
+    instant_launch_dashboard: dashboard
+    instant_launch_nav_drawer: nav_drawer
```
Am not entirely sure if these needs to be in config. I don't see them changing in different envs / external deployments.  They could live in some constants file and can be moved to configs when it is determined this will in fact be changing often. Also there is no real benefit to adding them in configs since we still have to redeploy the UI once the config changes ?

ash3rz replied:
Yeah, Paul made a similar comment about configs in another PR. It seems like we need to make a decision on how we want to handle them going forward.  It might be good to have @slr71 weigh in since she's been handling all the external deployments.

psarando replied:
I agree that these metadata attrs and values seem like they could be just integral to the DE system, and I'm not sure they would need to change between environments or deployments, like our system-use-only `ipc_*` iRODS metadata.

That said, maybe the metadata attributes could have a more specific name, in order to prevent any potential conflicts with future user defined metadata (like the `cyverse-community` attr above). So maybe `cyverse_de_ui_location`, for example?

In general I think configs are at least conceptually useful, and we should put values in configs that can (or should) change between environments or deployments, even if it requires a rebuild and redeploy for now. I just wish it were easier to change a config value in an environment and rolling-restart the ui service to pick up the new value, without having to push the change to a git repo and rebuild everything (especially if it's supposed to be just a temporary change, like disabling a feature for a little while).

psarando replied:
> That said, maybe the metadata attributes could have a more specific name, in order to prevent any potential conflicts with future user defined metadata (like the `cyverse-community` attr above). So maybe `cyverse_de_ui_location`, for example?
> 

Oh, I just noticed `ui_location` was already in use by older code, so scratch that suggestion 😅

slr71 replied:
Hmm, I’m not sure I see a reason to make that configurable. I suppose that language differences might mean that users from other countries would want to user terms that make sense to them, but probably only if they had to set the attributes directly. Since we’re planning to build a UI for it, I would think we could use constants instead.

ash3rz replied:
Cool - I'll refactor to use constants then.  Thanks for the input everybody!

sri2k1us reviewed (approved):
Code looks good to me. I had some thoughts on config changes and icon needs to updated.

ash3rz commented:
> Can we hard-code/ use some constants to get that icon in place for now until we get custom app icon feature implemented ?

Sure, we can hard-code it for now. I checked MUI's icons and didn't see anything that looked like a terminal but I saw you asked Mariah to make us one 👍

psarando commented:
> > Can we hard-code/ use some constants to get that icon in place for now until we get custom app icon feature implemented ?
> 
> Sure, we can hard-code it for now. I checked MUI's icons and didn't see anything that looked like a terminal but I saw you asked Mariah to make us one 👍

In the meantime, maybe another thing to try would be an `Avatar` with just some `>_` text?
```jsx
<Avatar><Typography variant="button">{">_"}</Typography></Avatar>
```

ash3rz commented:
Here's what I've got with Paul's idea:
![image](https://user-images.githubusercontent.com/8909156/135530955-93671730-45fb-4ddd-bc25-8cdac04ef1e0.png)
on hover:
![image](https://user-images.githubusercontent.com/8909156/135530974-028e35ad-6a07-46c1-9a1c-b4a02aef6f2b.png)

psarando commented:
> Here's what I've got with Paul's idea...

I was imagining the background color of the Avatar would be the lighter color and the text would be the darker color, like the `?` Help icon.

I guess that might require changing the color/backgroundColor CSS of the `Avatar`.

psarando commented on src/components/layout/AppBar.js:
```diff
@@ -239,6 +240,8 @@ function DEAppBar(props) {
         cacheTime: Infinity,
     });
 
+    const { instantLaunches } = useFirstClassInstantLaunch();
```
👍

psarando commented on src/components/instantlaunches/admin/InstantLaunchList.js:
```diff
@@ -149,7 +149,7 @@ const InstantLaunchList = ({ showErrorAnnouncer }) => {
     const dashboardILs = useQuery(
         [DASHBOARD_INSTANT_LAUNCHES_KEY, "ui_location", "dashboard"],
-        () => listInstantLaunchesByMetadata("ui_location", "dashboard")
+        () => adminListInstantLaunchesByMetadata("ui_location", "dashboard")
     );
```
or a constant, if we go with that instead...

psarando reviewed (approved):
LGTM 👍 ✨

ash3rz commented:
How about this?
![image](https://user-images.githubusercontent.com/8909156/135541908-e0de9b06-2916-4953-a30a-91d5cccec481.png)
There's also slightly rounded too:
![image](https://user-images.githubusercontent.com/8909156/135541943-a3aac8d8-87c3-47ff-b1b1-6a12a7eb6b4c.png)

psarando commented:
> There's also slightly rounded too

Nice, the rounded version LGTM 👍

ash3rz commented:
OK, I think I've got all the changes in.  The terminal icon is in and I refactored to use constants.  I also added a sign in dialog for when users click on the `cli` instant launch when not signed in.

ash3rz commented:
Here's Mariah's version with an svg: 
![image](https://user-images.githubusercontent.com/8909156/135550110-7333b7e6-d144-47c4-8765-55bfd28d3c0a.png)

psarando commented:
> Here's Mariah's version with an svg

Also LGTM 👍

ash3rz commented:
Thanks for the reviews!
